model/src/main/java/jetbrains/jetpad/model/composite/Composites.java

```diff
@@ -420,6 +433,82 @@ public void remove() {
     };
   }
 
+  private static <ValueT> Iterable<ValueT> iterateFrom(final ValueT initial, final Function<ValueT, ValueT> trans) {
+    return new Iterable<ValueT>() {
+      @Override
+      public Iterator<ValueT> iterator() {
+        return new Iterator<ValueT>() {
+          private ValueT myCurrent = initial;
+
+          @Override
+          public boolean hasNext() {
+            return myCurrent != null;
+          }
+
+          @Override
+          public ValueT next() {
+            if (myCurrent == null) {
+              throw new NoSuchElementException();
+            }
+            ValueT result = myCurrent;
+            myCurrent = trans.apply(myCurrent);
+            return result;
+          }
+
+          @Override
+          public void remove() {
+            throw new UnsupportedOperationException();
+          }
+        };
+      }
+    };
+  }
+
+  public static <CompositeT extends NavComposite<CompositeT>>
+      List<CompositeT> allBetween(CompositeT from, CompositeT to) {
+    List<CompositeT> res = new ArrayList<>();
+    res.add(from);
+
+    if (to != from) {
+      includeClosed(from, to, res);
+    }
+
+    return res;
+  }
+
+  private static <CompositeT extends NavComposite<CompositeT>>
+      boolean includeClosed(CompositeT left, CompositeT to, List<CompositeT> res) {
+    for (CompositeT next = left.nextSibling(); next != null; next = next.nextSibling()) {
+      if (includeOpen(next, to, res)) {
+        return true;
+      }
+    }
+
+    if (left.getParent() == null) {
```

Comment:
Not sure what should be asserted here. I don't see any problem with root in left.